### PR TITLE
Route away from non ready new-ui routes

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #4011 
# What's Changed

1. Instead of adding 'use_new_ui' to the Generic View class, which impacts all View Classes, I relocated it to each View class that is new ui ready.
2. Created a method that returns all of the new ui ready routes (regex).
3. To sync our frontend and backend, I established an API that would return all of the new UI ready routes, which would be used to disable the sidebar menus that are not ready.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by gsnider2195** [comment]
The issue #4011 is lacking any detail so I'm not sure if this was discussed but doesn't disabling the links in the nav menu kind of defeat the purpose of this change in a way? If we're going through the trouble of adding this feature, why not let the users click the link in the new ui nav menu and just send them to the legacy view?

**Comment by gsnider2195** [comment]
This doesn't appear to be working for me yet. If I click on a detail view for a location and then click on the link to the associated "Location Type" it sends me to the detail view for that location type in the new ui.

**Comment by timizuoebideri1** [comment]
> This doesn't appear to be working for me yet. If I click on a detail view for a location and then click on the link to the associated "Location Type" it sends me to the detail view for that location type in the new ui.

You should use the port `8080` as it only works for prod

**Comment by timizuoebideri1** [comment]
> The issue #4011 is lacking any detail so I'm not sure if this was discussed but doesn't disabling the links in the nav menu kind of defeat the purpose of this change in a way? If we're going through the trouble of adding this feature, why not let the users click the link in the new ui nav menu and just send them to the legacy view?

True, that makes sense

**Comment by timizuoebideri1** [comment]
> This doesn't appear to be working for me yet. If I click on a detail view for a location and then click on the link to the associated "Location Type" it sends me to the detail view for that location type in the new ui.

Wow, I actually forgot to implement that since we were disabling the nav menu. 😕

**Comment by gsnider2195** [comment]
There's a delay before redirecting to the old ui where the new ui is shown. I wonder if we could remove the routes for the unsupported views in the new UI and have a catch-all route that just reloads the page?

**Comment by timizuoebideri1** [comment]
> There's a delay before redirecting to the old ui where the new ui is shown. I wonder if we could remove the routes for the unsupported views in the new UI and have a catch-all route that just reloads the page?

They are not really different routes; we only have one route that accepts various combinations of model_name and app_label.

**Comment by timizuoebideri1** [comment]
> > There's a delay before redirecting to the old ui where the new ui is shown. I wonder if we could remove the routes for the unsupported views in the new UI and have a catch-all route that just reloads the page?
> 
> They are not really different routes; we only have one route that accepts various combinations of model_name and app_label.

We could replace the use of `Link` components with a custom component `NautobotLink` which can either use react routing `Link` or HTML `a` tag depending on if the `href` is newUI ready

**Comment by gsnider2195** [comment]
> > There's a delay before redirecting to the old ui where the new ui is shown. I wonder if we could remove the routes for the unsupported views in the new UI and have a catch-all route that just reloads the page?
> 
> They are not really different routes; we only have one route that accepts various combinations of model_name and app_label.

I'm proposing that we make them explicit routes. Instead of 
```
                        {
                            path: ":app_label/:model_name/",
                            element: <ListView />,
                            children: [],
                        },
```

We would do

```
                        {
                            path: "dcim/device-types/",
                            element: <ListView app_label="dcim" model_name="device-types" />,
                            children: [],
                        },
                        {
                            path: "dcim/devices/",
                            element: <ListView app_label="dcim" model_name="devices" />,
                            children: [],
                        },
                        {
                            path: "dcim/locations/",
                            element: <ListView app_label="dcim" model_name="locations" />,
                            children: [],
                        },
                        {
                            path: "ipam/ip-addresses/",
                            element: <ListView app_label="ipam" model_name="ip-addresses" />,
                            children: [],
                        },
```

**Comment by bryanculver** [comment]
> > > There's a delay before redirecting to the old ui where the new ui is shown. I wonder if we could remove the routes for the unsupported views in the new UI and have a catch-all route that just reloads the page?
> > 
> > 
> > They are not really different routes; we only have one route that accepts various combinations of model_name and app_label.
> 
> I'm proposing that we make them explicit routes. Instead of
> 
> ```
>                         {
>                             path: ":app_label/:model_name/",
>                             element: <ListView />,
>                             children: [],
>                         },
> ```
> 
> We would do
> 
> ```
>                         {
>                             path: "dcim/device-types/",
>                             element: <ListView app_label="dcim" model_name="device-types" />,
>                             children: [],
>                         },
>                         {
>                             path: "dcim/devices/",
>                             element: <ListView app_label="dcim" model_name="devices" />,
>                             children: [],
>                         },
>                         {
>                             path: "dcim/locations/",
>                             element: <ListView app_label="dcim" model_name="locations" />,
>                             children: [],
>                         },
>                         {
>                             path: "ipam/ip-addresses/",
>                             element: <ListView app_label="ipam" model_name="ip-addresses" />,
>                             children: [],
>                         },
> ```

This might make it harder to develop forward but maybe `is dev` check to allow us the previous behavior as we polish more views.

**Comment by gsnider2195** [comment]
If you navigate to a view in the legacy ui that's not ready for the new ui and you click on the "View in New UI" link it does nothing. It just appears broken. Perhaps we need to alert the user that the noop was intentional?

**Comment by gsnider2195** [comment]
Can you mark the login view as new ui ready?

File: changes/4011.added [hunk 1]
```diff
@@ -0,0 +1 @@
+Redirect user in new UI to legacy UI if the page being visited is not new UI ready.
```
Comment (gsnider2195) [inline changes/4011.added:1]:
```suggestion
Added redirect in new UI to legacy UI if the page being visited is not new UI ready.
```

File: nautobot/core/utils/navigation.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                # Check if the current action is part of the allowed actions in this ViewSet class
+                use_new_ui = bool(set(use_new_ui_list) & set(pattern.callback.actions.values()))
+            if use_new_ui:
+                r_pattern = pattern.pattern.regex.pattern.lstrip("^")
+                final_pattern = rf"^{prefix}{r_pattern}"
+                new_ui_routes.add(final_pattern)
+    return new_ui_routes
+
+
+def get_all_new_ui_ready_route():
```
Comment (gsnider2195) [inline nautobot/core/utils/navigation.py:43]:
```suggestion
def get_all_new_ui_ready_routes():
```

File: nautobot/circuits/views.py [hunk 3]
```diff
@@ -135,6 +135,8 @@ class CircuitUIViewSet(NautobotUIViewSet):
     queryset = Circuit.objects.all()
     serializer_class = serializers.CircuitSerializer
     table_class = tables.CircuitTable
+    # NOTE: This is how `NautobotUIViewSet` would define use_new_ui attr
+    # use_new_ui = ["list", "retrieve"]
```
Comment (gsnider2195) [inline nautobot/circuits/views.py:139]: Why only here? We're using NautobotUIViewSet in a bunch of views:

```
nautobot/dcim/views.py:2815:class DeviceRedundancyGroupUIViewSet(NautobotUIViewSet):
nautobot/dcim/views.py:2839:class InterfaceRedundancyGroupUIViewSet(NautobotUIViewSet):
nautobot/circuits/views.py:97:class ProviderUIViewSet(NautobotUIViewSet):
nautobot/circuits/views.py:128:class CircuitUIViewSet(NautobotUIViewSet):
nautobot/circuits/views.py:181:class ProviderNetworkUIViewSet(NautobotUIViewSet):
nautobot/extras/views.py:1518:class JobButtonUIViewSet(NautobotUIViewSet):
nautobot/extras/views.py:1777:class RoleUIViewSet(viewsets.NautobotUIViewSet):
```

File: nautobot/core/utils/navigation.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import re
+from django.urls import get_resolver
```
Comment (gsnider2195) [inline nautobot/core/utils/navigation.py:2]:
```suggestion

from django.urls import get_resolver
```

**Review by gsnider2195 — APPROVED** [review]
Just some minor nitpicks, otherwise looks good!

File: nautobot/circuits/views.py (hunk 3 above)
Comment (timizuoebideri1) [reply nautobot/circuits/views.py:139]: It's random, and it's simply a note to users/anyone planning on getting NautobotUIViewSet new_ui ready on how to implement it.

**Pull request merged**